### PR TITLE
Animation Lab: Animate does not take into account shape flipping #1075

### LINKED ISSUE #1075: Animation Lab: Animate does not take into account shape flipping
This problem only occurs when shapes are modified using the flip tool or resized such that it flips. It is present on both `Add Animation Slide` and `Animate In Slide`.

**Current Behaviour**

actual

**Expected Behaviour**

expected

### PR DESCRIPTION
**Pull request by jamos-tay**
Fixes #1075.

The 'expected behaviour' wouldn't work because the arrow ends up upside down rather than flipped horizontally. Rather, we should animate the horizontal flip.

New behaviour:

expected

Flipped vertical:

vert

Flipped horizontal:

horiz

Flipped both ways:

both

### CODE REVIEW

**Comment by kai33** [comment]
@damithc @Ohohcakester  is this the expected behaviour?

**Comment by damithc** [comment]
@kai33 yes, it is an improvement over the current behavior. 'Flipped bothways' looks a bit odd, but still better than the current behavior.

File: PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs [hunk 1]
```diff
@@ -30,6 +30,17 @@ public static void AddFrameMotionAnimation(PowerPointSlide animationSlide, Power
             float finalHeight = finalShape.Height;
             float finalFont = 0.0f;
 
+            bool flipHorizontal = initialShape.HorizontalFlip != finalShape.HorizontalFlip;
+            bool flipVertical = initialShape.VerticalFlip != finalShape.VerticalFlip;
```
Comment (kai33) [inline PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs]: bool start with `is`?

File: PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs [hunk 2]
```diff
@@ -92,10 +103,10 @@ public static void AddZoomToAreaPanFrameMotionAnimation(PowerPointSlide animatio
             float incrementLeft = (finalX - initialX) / numFrames;
             float incrementTop = (finalY - initialY) / numFrames;
 
-            AddFrameAnimationEffects(animationSlide, initialShape, incrementLeft, incrementTop, incrementWidth, incrementHeight, 0.0f, 0.0f, duration, numFrames);
+            AddFrameAnimationEffects(animationSlide, initialShape, incrementLeft, incrementTop, incrementWidth, incrementHeight, false, false, 0.0f, 0.0f, duration, numFrames);
         }
 
-        private static void AddFrameAnimationEffects(PowerPointSlide animationSlide, PowerPoint.Shape initialShape, float incrementLeft, float incrementTop, float incrementWidth, float incrementHeight, float incrementRotation, float incrementFont, float duration, int numFrames)
+        private static void AddFrameAnimationEffects(PowerPointSlide animationSlide, PowerPoint.Shape initialShape, float incrementLeft, float incrementTop, float incrementWidth, float incrementHeight, bool flipHorizontal, bool flipVertical, float incrementRotation, float incrementFont, float duration, int numFrames)
```
Comment (kai33) [inline PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs]: function param, boolean starts with `is`

Comment (kai33) [inline PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs]: like isFlipHorizontally

File: PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs [hunk 3]
```diff
@@ -92,10 +103,10 @@ public static void AddZoomToAreaPanFrameMotionAnimation(PowerPointSlide animatio
             float incrementLeft = (finalX - initialX) / numFrames;
             float incrementTop = (finalY - initialY) / numFrames;
 
-            AddFrameAnimationEffects(animationSlide, initialShape, incrementLeft, incrementTop, incrementWidth, incrementHeight, 0.0f, 0.0f, duration, numFrames);
+            AddFrameAnimationEffects(animationSlide, initialShape, incrementLeft, incrementTop, incrementWidth, incrementHeight, false, false, 0.0f, 0.0f, duration, numFrames);
```
Comment (kai33) [inline PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs]: add a prefix of param name to indicate the meaning of the boolean value,
e.g. if the param name is isFlipHorizontally, then `isFlipHorizontally: false`

**Comment by kai33** [comment]
Hi Jamos thanks for this fix.
Can you also create a functional test to guard your changes?

**Comment by Ohohcakester** [comment]
Yeah, the behaviour is expected behaviour to me.
I see scaling as a generalisation of flipping. (scale by -1 along an axis)

File: PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs (hunk 3 above)
Comment (jamos-tay) [inline PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs]: If I add the prefix I must also add them for the other 4 params, is that okay?

image

**Comment by jamos-tay** [comment]
Okay, will do.

Also, @damithc suggested that for flipping both ways, the scaling should be skewed differently for both axes so it doesn't shrink to a dot. I'll try to implement that too.

File: PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs (hunk 3 above)
Comment (kai33) [inline PowerPointLabs/PowerPointLabs/FrameMotionAnimation.cs]: it's fine

**Comment by kai33** [comment]
Hi Jamos, just rebase this pr and ensure FT still passes, then I can merge it

**Comment by jamos-tay** [comment]
Hi, should be good to go

**Pull request merged**